File: typescript/ch5/MonadWrapper.ts
import { WrapperFunc } from './WrapperFunc';

class MonadWrapper {
  protected _value: any;

  static of(value: any) {
    return new MonadWrapper(value);
  }

  constructor(value: any) {
    this._value = value;
  }

  map(f: WrapperFunc) {
    return MonadWrapper.of(f(this._value));
  }

  join(): MonadWrapper {
    if (!(this._value instanceof MonadWrapper)) {
      return this;
    }

    return this._value.join();
  }

  get() {
    return this._value;
  }

  toString() {
    return `Wrapper(${this._value})`;
  }
}

export {
  MonadWrapper
};

File: python/python/ch5/MonadWrapper.py
class MonadWrapper:
    def __init__(self, value):
        self._value = value
        
    @staticmethod
    def of(value):
        return MonadWrapper(value)
      
    def map(self, f):
        return MonadWrapper.of(f(self._value))
      
    def join(self):
        if not isinstance(self._value, MonadWrapper):
            return self
          
        return self._value.join()
      
    def get(self):
        return self._value
      
    def __str__(self):
        return f'Wrapper({self._value})'